### PR TITLE
Determine slice number in L0 generation (Issue 13)

### LINKED ISSUE #13: Missing slice number from generated output file
I'm trying to simulate L0 step 2 generation with the scenario configuration and working directory defined in [L0PFS2_WD_scenario.zip](https://github.com/stcorp/procsim/files/7538214/L0PFS2_WD_scenario.zip).

In the L0 step2 generation, the CPF provides to the processor the input data of the slice _n_ plus the previous and next slices. In this case, see job order BIO_L0PFS2_01.00_7217244-03_20211114T120000_JobOrder.xml, the slice to be processed is the slice 3 (2017-03-02T06:08:56.546000, 2017-03-02T06:09:37.672000), i.e. the last one in the data take, thus, only the input data related to the previous slice are passed to the processor:

```xml
<Selected_Input>
  <File_Type>RAWS025_10</File_Type>
  <List_of_File_Names>
    <File_Name>/opt/pnode/data/WD-L0PFS2-7217244-03/inputs/BIO_RAWS025_10_20170302T060723_20170302T060908_D20170302T073617_01_BEU0LE</File_Name>
    <File_Name>/opt/pnode/data/WD-L0PFS2-7217244-03/inputs/BIO_RAWS025_10_20170302T060856_20170302T060937_D20170302T073617_01_BEU0LE</File_Name>
  </List_of_File_Names>
</Selected_Input>
```

Procsim correctly selects scenario name `LOAD_TDS - SM idx 0 DTID 7217244 L0PFS2` (see `l0-scenarios.json` attached).

The generated output products are:

```
BIO_S3_RAWP_0M_20170302T060856_20170302T060937_I_G01_M02_C03_T001_F____01_BEZB7C
BIO_S3_RAW__0S_20170302T060856_20170302T060937_I_G01_M02_C03_T001_F____01_BEZB7C
```

The frame field in the file name is set to `F___` instead of `F003`.

The `eop:wrsLatitudeGrid` parameter, in the MPH, is set to:

```xml
<eop:wrsLatitudeGrid eop:codeSpace="urn:esa:eop:Biomass:frames">___</eop:wrsLatitudeGrid>
```

instead of:

```xml
<eop:wrsLatitudeGrid eop:codeSpace="urn:esa:eop:Biomass:frames">3</eop:wrsLatitudeGrid>
```

To solve the issue, I have added the following parameter: `slice_frame_nr=3` in the `LOAD_TDS - SM idx 0 DTID 7217244 L0PFS2` but, the result is that Procsim always assign the value 3 to the output products, even if I'm trying to process slice 2.

Creating a different scenario for each slice to be processed does not solve the issue because Procsim always selects the first applicable scenario found. E.g.:

```
{
    "name": "LOAD_TDS - SM idx 0 DTID 7217244 L0PFS2 slice 2",
    "file_name": "l0pfs2.sh",
    "processor_name": "L0PFS2",
    "processor_version": "01.00",
    "task_name": "L0PFS2",
    "task_version": "01.00",
    "log_level": "debug",
    "data_takes": [
        {
            "data_take_id": 7217244,
            "start": "2017-03-02T06:07:23.674000Z",
            "stop": "2017-03-02T06:09:37.672000Z",
            "swath": "S3",
            "operational_mode": "SM"
        }
    ],
    "begin_end_position_from_toi": true,
    "mission_phase": "Interferometric",
    "global_coverage_id": "01",
    "major_cycle_id": "02",
    "repeat_cycle_id": "03",
    "track_nr": "1",
    "slice_frame_nr": 2,
    "num_l0_lines": "314841",
    "num_l0_lines_corrupt": "0",
    "num_l0_lines_missing": "0",
    "outputs": [
        {
            "type": "S3_RAW__0S"
        },
        {
            "type": "S3_RAWP_0M"
        }
    ]
},
{
    "name": "LOAD_TDS - SM idx 0 DTID 7217244 L0PFS2 slice 2",
    "file_name": "l0pfs2.sh",
    "processor_name": "L0PFS2",
    "processor_version": "01.00",
    "task_name": "L0PFS2",
    "task_version": "01.00",
    "log_level": "debug",
    "data_takes": [
        {
            "data_take_id": 7217244,
            "start": "2017-03-02T06:07:23.674000Z",
            "stop": "2017-03-02T06:09:37.672000Z",
            "swath": "S3",
            "operational_mode": "SM"
        }
    ],
    "begin_end_position_from_toi": true,
    "mission_phase": "Interferometric",
    "global_coverage_id": "01",
    "major_cycle_id": "02",
    "repeat_cycle_id": "03",
    "track_nr": "1",
    "slice_frame_nr": 3,
    "num_l0_lines": "314841",
    "num_l0_lines_corrupt": "0",
    "num_l0_lines_missing": "0",
    "outputs": [
        {
            "type": "S3_RAW__0S"
        },
        {
            "type": "S3_RAWP_0M"
        }
    ]
}
```

I think that the only way to handle this situation is by letting Procsim to calculate the slice grid and number also for the generation of Sx_RAW__0S, Sx_RAWP_0M, EC_RAWP_0x, as already done for the RAWS0xx_10 products using the anx information provided in the simulation.

Do you agree? Do you have other solutions?

### PR DESCRIPTION
**Pull request by mivermeu**
This PR aims to resolve #13 by generalising ANX and frame/slice number calculations. Previously they were only used in the raw product generator. Now they are housed in the general product generator for use anywhere, although they are not required to be set/used.

### CODE REVIEW

File: procsim/biomass/level0_product_generator.py [hunk 1]
```diff
@@ -317,8 +324,10 @@ def generate_output(self):
         self._hdr.product_type = self._resolve_wildcard_product_type()
         self._hdr.incomplete_l0_slice = False
         self._hdr.partial_l0_slice = False
-        acq = self._hdr.acquisitions[0]
-        acq.slice_frame_nr = None
+        
+        # Get slice number from middle of slice to deal with merged slices.
+        middle_position = self._hdr.begin_position + (self._hdr.end_position - self._hdr.begin_position) / 2
+        self._hdr.acquisitions[0].slice_frame_nr = self._get_slice_frame_nr(middle_position, constants.SLICE_GRID_SPACING)
```
Comment (pe1obw) [inline procsim/biomass/level0_product_generator.py:330]: I think this is wrong, according to the class description: "set wrsLatitudeGrid, aka the slice_frame_nr, to '___'".
I assumed that these merged products have no slice number. Or did I miss something?

File: procsim/biomass/level0_product_generator.py [hunk 2]
```diff
@@ -408,8 +417,11 @@ def generate_output(self):
         self._hdr.product_type = self._output_type
         self._hdr.incomplete_l0_slice = False
         self._hdr.partial_l0_slice = False
-        acq = self._hdr.acquisitions[0]
-        acq.slice_frame_nr = None
+        
+        # Get slice number from middle of slice to deal with merged slices.
+        middle_position = self._hdr.begin_position + (self._hdr.end_position - self._hdr.begin_position) / 2
+        self._hdr.acquisitions[0].slice_frame_nr = self._get_slice_frame_nr(middle_position, constants.SLICE_GRID_SPACING)
```
Comment (pe1obw) [inline procsim/biomass/level0_product_generator.py:423]: See previous comment

File: procsim/biomass/level0_product_generator.py [hunk 3]
```diff
@@ -165,6 +166,12 @@ def _generate_product(self, start, stop, data_take_config):
             start,
             stop
         )
+            
+        # Determine and set the slice number if not set already.
+        if self._hdr.acquisitions[0].slice_frame_nr is None:
+            # Get slice number from middle of slice to deal with merged slices.
+            middle = start + (stop - start) / 2
```
[the remark this answers is not in the document]
Comment (mivermeu) [reply procsim/biomass/level0_product_generator.py:173]: I've created issue #20 to address this.

File: procsim/biomass/level0_product_generator.py (hunk 1 above)
Comment (mivermeu) [reply procsim/biomass/level0_product_generator.py]: I think I was the one who missed something when I assumed these products were sliced. I've reverted it in https://github.com/stcorp/procsim/pull/17/commits/672e56460cacbfd00ebf94ab26b392ad0791c5a1.

File: procsim/biomass/level0_product_generator.py (hunk 2 above)
Comment (mivermeu) [reply procsim/biomass/level0_product_generator.py]: Reverted in https://github.com/stcorp/procsim/pull/17/commits/672e56460cacbfd00ebf94ab26b392ad0791c5a1.

**Pull request merged**